### PR TITLE
[None][feat] Add Tencent HunYuanDenseV1 model support

### PR DESCRIPTION
**Pull request by sorenwu**
## Summary by CodeRabbit

- New Features
  - Added support for configuring and running a HunYuan dense causal language model, with options for attention variants, KV heads, rotary embeddings, and normalization choices.
  - Publicly exposed the new model for easy use via the standard API.
  - Introduced optional decode-time debugging to dump per-layer tensor statistics for troubleshooting.

## Description

Currently, the Hunyuan inference team supports the Hunyuan-0.5B/1.8B/4B/7B model. By adding the modeling_hunyuan_dense.py related files, it supports the model of HunYuanDenseV1ForCausalLM.

We have validated the accuracy of this PR，HunYuan (new Dense LLM model from Tencent) will open source these days.

Thanks~

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/model_config.py [hunk 1]
```diff
@@ -23,6 +24,368 @@
 TConfig = TypeVar("TConfig", bound=transformers.PretrainedConfig)
 
 
+class HunYuanPretrainedConfig(PretrainedConfig):
```
Comment (syuoni) [inline tensorrt_llm/_torch/model_config.py:27]: Does `HunYuanPretrainedConfig` supported by HF transformers? 
* If so, this config definition is not necessary
* Otherwise, please move it to `modeling_hunyuan_dense.py`

File: tensorrt_llm/_torch/model_config.py [hunk 2]
```diff
@@ -23,6 +24,368 @@
...
+        # if 'rope_scaling' in model_config and model_config['rope_scaling']:
+        #     rope_scaling = model_config['rope_scaling']
+        #     scaling_type = rope_scaling['type'] if 'type' in rope_scaling else 'dynamic'
+        #     assert scaling_type == 'dynamic', f"Now only support dynamic rope scaling, but got {scaling_type}"
+        #     if 'alpha' in rope_scaling and rope_scaling['alpha']:
+        #         scaling_alpha = rope_scaling['alpha']
+        #     else:
+        #         scaling_alpha = 1.0
+
+        #     rope_base = rope_theta * scaling_alpha ** (size_per_head / (size_per_head - 2))
```
Comment (syuoni) [inline tensorrt_llm/_torch/model_config.py:355]: Please clean the code.

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+from ..modules.linear import TensorParallelMode
+from ..modules.multi_stream_utils import maybe_execute_in_parallel
+from ..modules.rms_norm import RMSNorm
+from ..modules.rotary_embedding import RotaryEmbedding
+from ..utils import AuxStreamType, Fp4QuantizedTensor
+from .modeling_utils import (DecoderModel, DecoderModelForCausalLM,
+                             duplicate_kv_weight, register_auto_model)
+
+
+def _get_tensor_stats(tensor):
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:29]: This is for debug use, right? Please clean up such kinds of code.

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+
+    return first_values, middle_values, last_values, mean, var
+
+
+class HunYuanTRTRotaryEmbedding(RotaryEmbedding):
+
+    def __init__(self,
+                 config: HunYuanPretrainedConfig,
+                 device: Optional[torch.device] = None):
+        "default" if config.rope_scaling is None else 'hunyuan'
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:80]: unused code line.

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+            mrope_config=mrope_config,
+            all_reduce_params=all_reduce_params,
+            lora_params=lora_params,
+            **kwargs,
+        )
+
+
+class HunYuanMLP(GatedMLP):
+    """
+    继承自 GatedMLP 的 HunYuanMLP 实现, 兼容混合专家和tp
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:192]: To be consistent with other code files in this repo, please use English for doc and comments, thanks!

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+                         bias=getattr(model_config, "mlp_bias", False),
+                         activation=activation,
+                         dtype=getattr(model_config, "torch_dtype",
+                                       torch.bfloat16),
+                         config=model_config,
+                         overridden_tp_size=1,
+                         is_expert=is_moe)
+
+    def forward(self, x):
+        return super().forward(x)
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:232]: Seems this forward can be removed.

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+            attn_metadata=attn_metadata,
+            attention_mask=attention_mask,
+            mrope_config=mrope_config,
+            all_reduce_params=all_reduce_params,
+            lora_params=lora_params,
+            **kwargs,
+        )
+
+
+class HunYuanMLP(GatedMLP):
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:190]: Seems that `HunYuanMLP` is ununsed?

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+            position_ids=position_ids,
+            hidden_states=hidden_states,
+            attn_metadata=attn_metadata,
+            **kwargs,
+        )
+        # Fully Connected
+        hidden_states = residual + hidden_states
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        hidden_states = self.mlp(hidden_states, is_hunyuan=True)
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:294]: The `self.mlp` is an instance of `GatedMLP`, it should not accept `is_hunyuan=True`

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+    def forward(
+        self,
+        attn_metadata: AttentionMetadata,
+        input_ids: Optional[torch.LongTensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        return_context_logits: bool = False,
+        **kwargs,
+    ) -> torch.Tensor:
+        ###### DEBUG #####
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:438]: Please clean up the debug-related code.

**Review by syuoni — COMMENTED** [review]
Hi @sorenwu , thanks for the contribution! Please see the comments.

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py (hunk 9 above)
Comment (sorenwu) [reply tensorrt_llm/_torch/models/modeling_hunyuan_dense.py]: Ok, I have edited the question above

File: tensorrt_llm/_torch/models/modeling_hunyuan_dense.py [hunk 10]
```diff
@@ -0,0 +1,750 @@
...
+    except Exception as e:
+        # 如果计算失败，返回None
+        print(
+            f"Warning: Could not compute statistics for tensor with dtype {flat.dtype}: {e}"
+        )
+        mean, var = None, None
+
+    return first_values, middle_values, last_values, mean, var
+
+
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_hunyuan_dense.py:420]:
```suggestion
```

**Review by syuoni — APPROVED** [review]
LGYM, thanks!

**Comment by syuoni** [comment]
This model requires transformers>=4.56.0, waiting for https://github.com/NVIDIA/TensorRT-LLM/pull/7523
